src/f32-gemm/MrxLmulv-rvv.c.in

```diff
@@ -0,0 +1,170 @@
+// Copyright 2023 SiFive, Inc.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+$assert DATATYPE in ["F32"]
+$assert ACTIVATION in ["LINEAR", "RELU", "MINMAX"]
+$assert MR >= 1
+$assert LMUL in [1, 2, 4]
+#include <assert.h>
+
+#include <riscv_vector.h>
+
+#include <xnnpack/gemm.h>
+
+
+$DATATYPE_SPEC = {"F32": "f32"}[DATATYPE]
+$SUFFIX = {"LINEAR": "", "RELU": "_relu", "MINMAX": "_minmax"}[ACTIVATION]
+$PARAMS = {"LINEAR": "xnn_f32_default_params", "RELU": "xnn_f32_relu_params", "MINMAX": "xnn_f32_minmax_params"}[ACTIVATION]
+void xnn_${DATATYPE_SPEC}_gemm${SUFFIX}_ukernel_${MR}x${LMUL}v__rvv(
+    size_t mr,
+    size_t nc,
+    size_t kc,
+    const float* restrict a,
+    size_t a_stride,
+    const float* restrict w,
+    float* restrict c,
+    size_t cm_stride,
+    size_t cn_stride,
+    const union ${PARAMS} params[restrict XNN_MIN_ELEMENTS(1)])
+{
+  assert(mr != 0);
+  assert(mr <= ${MR});
+  assert(nc != 0);
+  assert(kc != 0);
+  assert(kc % sizeof(float) == 0);
+  assert(a != NULL);
+  assert(w != NULL);
+  assert(c != NULL);
+
+  $if ACTIVATION == "MINMAX":
+    const float vmin = params->scalar.min;
+    const float vmax = params->scalar.max;
+  const float* a0 = a;
+  float* c0 = c;
+  $if MR > 1:
+    const float* a1 = (const float*) ((uintptr_t) a + a_stride);
+    float* c1 = (float*) ((uintptr_t) c + cm_stride);
+    $for M in range(2, MR):
+      const float* a${M} = (const float*) ((uintptr_t) a${M-1} + a_stride);
+      float* c${M} = (float*) ((uintptr_t) c${M-1} + cm_stride);
+
+  const size_t nr = __riscv_vsetvlmax_e32m${LMUL}();
+  size_t vl = nr;
+  if XNN_LIKELY(mr == ${MR}) {
+    do {
+      if XNN_UNLIKELY(nc < nr) {
+        vl = __riscv_vsetvl_e32m${LMUL}(nc);
+      }
+      nc = nc - vl;
+
+      vfloat32m${LMUL}_t v_acc0 =  __riscv_vle32_v_f32m${LMUL}(w, vl);
+      w = w + nr;
+      $for M in range(1, MR):  
+        vfloat32m${LMUL}_t v_acc${M} =  __riscv_vmv_v_v_f32m${LMUL}(v_acc0, vl);
+
+      size_t k = kc;
+      do {
+        vfloat32m${LMUL}_t v_w_k = __riscv_vle32_v_f32m${LMUL}(w, vl);
+        w = w + nr;
+        $for M in range(0, MR):  
+          v_acc${M} = __riscv_vfmacc_vf_f32m${LMUL}(v_acc${M}, *a${M}, v_w_k, vl);
+          a${M}++;
+        k -= sizeof(float);
```

Comment:
doing 1 at a time likely isnt the most efficient?
With 8 bit its even less efficient to just do 1 byte, and the best approach tends to be KR=4
There are many approaches
lanes - arm used 2 or 4 floats and lanes to select which to multiply against
dup - intel doesnt have lanes, so it can use a shuffle instruction to isolate 1 float from a vector
shuffle - read a full vector of input, multiply by weights, and then rotate it by 1 float.
kr - read KR floats of input and outside the loop, horizontally add the KR accumulators.  Or use a dot product inside the loop.  KR is also how matrix and dotproduct instructions are implemented.
acc2 - unroll the loop by 2 and create 2 sets of accumulators.

Is there anything on riscv that would allow 2 or 4 floats of input at a time?

Reply:
> doing 1 at a time likely isnt the most efficient? With 8 bit its even less efficient to just do 1 byte, and the best approach tends to be KR=4
> There are many approaches lanes - arm used 2 or 4 floats and lanes to select which to multiply against dup - intel doesnt have lanes, so it can use a shuffle instruction to isolate 1 float from a vector shuffle - read a full vector of input, multiply by weights, and then rotate it by 1 float. 
kr - read KR floats of input and outside the loop, horizontally add the KR accumulators. Or use a dot product inside the loop. KR is also how matrix and dotproduct instructions are implemented. acc2 - unroll the loop by 2 and create 2 sets of accumulators.

I think that set kr >= 1 might lead to poor performance. RVV v-spec doesn't have basic instructions to align to neon dot-product instructions.

For instance, if `kr=2, k =3, m=1 nr=4`:

```
Load a0, a1 with vl=2.
va0 = a00 | a01 

Load weight contains:
set vl to `nr * kr`
vb = w00 | w01 | w10 | w11 | w20 | w21 | w30 | W31
```
To get the partial sum, duplicate va0 by vrgather
(we can imagine it might be slow).
```
va0 = a00 | a01 | a00 | a01 | a00 | a01 |  a00 | a01
multiply va0 & vb 
vacc = w00 a00 | w01 a01 | ....
```
After going through all k, we need to pair add to get the accumulator results. However, there is no pair-add in risc-v vector...
You need to use several instructions to imitate.

> Is there anything on riscv that would allow 2 or 4 floats of input at a time?

You need to change vl to 2~4 and then load it to v_reg, so it might not be efficient.

On the other side, I think acc=2 might be a good method to try for the case that MR is small(MR=1).
However, is there any use case for Mr = 1?
Based on my understanding, specializing for Mr=1 isn't necessary for GEMM, but IGEMM is necessary in gemm-config.